## Аналіз A/B тесту

Маємо проаналізувати дані A/B тесту в популярній [грі Cookie Cats](https://www.facebook.com/cookiecatsgame). Це класична гра-головоломка в стилі «з’єднай три», де гравець повинен з’єднати плитки одного кольору, щоб очистити дошку та виграти рівень. На дошці також зображені співаючі котики :)

Під час проходження гри гравці стикаються з воротами, які змушують їх чекати деякий час, перш ніж вони зможуть прогресувати або зробити покупку в додатку. У цьому блоці завдань ми проаналізуємо результати A/B тесту, коли перші ворота в Cookie Cats було переміщено з рівня 30 на рівень 40. Зокрема, ми проаналізуємо вплив на утримання (retention) гравців. Тобто хочемо зрозуміти чи переміщення воріт на 10 рівнів пізніше якимось чином вплинуло на те, що користувачі перестають грати в гру раніше чи пізніше з точки зору кількості їх днів з моменту встановлення гри.

Будемо працювати з даними з файлу `cookie_cats.csv`. Змінні в даних наступні:

- userid - унікальний номер, який ідентифікує кожного гравця.
- version - чи потрапив гравець в контрольну групу (gate_30 - ворота на 30 рівні) чи тестову групу (gate_40 - ворота на 40 рівні).
- sum_gamerounds - кількість ігрових раундів, зіграних гравцем протягом першого тижня після встановлення
- retention_1 - чи через 1 день після встановлення гравець повернувся і почав грати?
- retention_7 - чи через 7 днів після встановлення гравець повернувся і почав грати?

Коли гравець встановлював гру, його випадковим чином призначали до групи gate_30 або gate_40.

1. Зчитайте дані АВ тесту у змінну `df` та виведіть середнє значення показника показник `retention_7` (утримання на 7 день) по версіям гри. Сформулюйте гіпотезу: яка версія дає краще утримання через 7 днів після встановлення гри?

Гіпотеза: переміщення воріт з рівня 30 на рівень 40 позитивно вплине на утримання гравців протягом 7 днів  
Мета: зрозуміти, чи є позитивна кореляція між рівнем розташування воріт та тривалістю гри юзерів з моменту встановлення

In [28]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import statsmodels.stats.api as sms
from math import ceil

%matplotlib inline

In [2]:
df = pd.read_csv('../data/cookie_cats.csv')

df.head()

userid  version  sum_gamerounds  retention_1  retention_7
0     116  gate_30               3        False        False
1     337  gate_30              38         True        False
2     377  gate_40             165         True        False
3     483  gate_40               1        False        False
4     488  gate_40             179         True         True

In [4]:
mean_retention_by_version = df.groupby('version')['retention_7'].mean()

mean_retention_by_version

version
gate_30    0.190201
gate_40    0.182000
Name: retention_7, dtype: float64

In [12]:
users_count = df.groupby('version')['userid'].count()

users_count

version
gate_30    44700
gate_40    45489
Name: userid, dtype: int64

In [13]:
required_n = users_count.min()
required_n

44700

`gate_30`: Гравці у цій версії гри мають трохи краще утримання через 7 днів, ніж у версії gate_40.
Можна припустити, що розташування перших "воріт" на 30-му рівні впливає позитивніше на гравців, ніж на 40-му рівні.

2. Перевірте з допомогою z-тесту аналогічно до прикладу в лекції, чи дає якась з версій гри кращий показник `retention_7` на рівні значущості 0.05. Обчисліть також довірчі інтервали для двох вибірок. Виведіть результат у форматі:
```
z statistic: ...
p-value: ...
Довірчий інтервал 95% для групи control: [..., ...]
Довірчий інтервал 95% для групи treatment: [..., ...]
```
де замість `...` - обчислені значення. В якості висновка дайте відповідь на два питання:  
    1. чи є статистична значущою різниця між поведінкою користувачів у різних версіях гри?   
    2. чи перетинаються довірчі інтервали утримання користувачів з різних версій гри? Про що це каже?  
    
Зверніть увагу, в такому і схожому завданнях ми використовуєм `proportion` Z-тест. Це тому що в нас залежна змінна має бінарне значення (повернеться аби ні користувач, чи клікне або ні користувач в інших ситуаціях - всього два можливих значення в змінної: 0/1, True/False ). Якщо б ми вимірювали скажімо чи є стат. значущою різниця між вагою чоловіків і жінок в певній вибірці, ми б використовувавли функцію `statsmodels.stats.ztest`, бо залежна змінна `вага` є неперервною (тип float, замість типу int чи bool і тільки двох можливих значень).

In [ ]:
control_sample = df[df['version'] == 'gate_30'].sample(n=required_n, random_state=22)
treatment_sample = df[df['version'] == 'gate_40'].sample(n=required_n, random_state=22)

ab_test = pd.concat([control_sample, treatment_sample], axis=0)
ab_test.reset_index(drop=True, inplace=True)

ab_test

userid  version  sum_gamerounds  retention_1  retention_7
0      7540471  gate_30              45         True        False
1      3589138  gate_30              21         True        False
2      3177668  gate_30              14         True        False
3      2133884  gate_30              26        False        False
4       492763  gate_30              39         True         True
...        ...      ...             ...          ...          ...
89395  3795601  gate_40              35         True        False
89396  8984080  gate_40             136         True        False
89397  3853652  gate_40              66        False         True
89398   772383  gate_40               3        False        False
89399  8331655  gate_40              24        False        False

[89400 rows x 5 columns]

In [15]:
ab_test['version'].value_counts()

version
gate_30    44700
gate_40    44700
Name: count, dtype: int64

In [18]:
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

control_results = ab_test[ab_test['version'] == 'gate_30']['retention_7']
treatment_results = ab_test[ab_test['version'] == 'gate_40']['retention_7']

In [20]:
n_con = control_results.count()
n_treat = treatment_results.count()
successes = [control_results.sum(), treatment_results.sum()]
nobs = [n_con, n_treat]

z_stat, p_value = proportions_ztest(successes, nobs=nobs)
(lower_con, lower_treat), (upper_con, upper_treat) = proportion_confint(successes, nobs=nobs, alpha=0.05)

print(f'z statistic: {z_stat:.2f}')
print(f'p-value: {p_value:.3f}')
print(f'Довірчий інтервал 95% для групи control: [{lower_con:.3f}, {upper_con:.3f}]')
print(f'Довірчий інтервал 95% для групи treatment: [{lower_treat:.3f}, {upper_treat:.3f}]')


z statistic: 3.19
p-value: 0.001
Довірчий інтервал 95% для групи control: [0.187, 0.194]
Довірчий інтервал 95% для групи treatment: [0.178, 0.185]


1. Чи є статистично значущою різниця між поведінкою користувачів у різних версіях гри?  
Так, є. Значення z-тесту становить 3.19, а p-value дорівнює 0.001, що є значно меншим за стандартний рівень значущості 0.05. Це свідчить про те, що ми відхиляємо нульову гіпотезу і що різниця між поведінкою користувачів у різних версіях гри є статистично значущою.

2. Чи перетинаються довірчі інтервали утримання користувачів з різних версій гри? Про що це каже?  
Довірчі інтервали не перетинаються, що також свідчить про наявність суттєвої різниці між утриманням користувачів у двох групах. Відсутність перетинів явно вказує на те, що версія гри з воротами на рівні 30 демонструє краще утримання користувачів на 7-й день, ніж версія з воротами на рівні 40.

3. Є ще один тип тестів, який використовується для бінарної метрики як от "зробить юзер дію, чи ні" - тест [**Хі-квадрат**](https://www.bmj.com/about-bmj/resources-readers/publications/statistics-square-one/8-chi-squared-tests) (ще одне [пояснення](https://www.scribbr.com/statistics/chi-square-tests/) тесту з прикладами). В нього інші гіпотези Н0 і Н1 на відміну від z- та t-тестів. А також цей тест можна використовувати, якщо в нас більше за 2 досліджувані групи, тобто в нас не А/В тест, а А/B/C/D, наприклад.  

В **z- та t-тестах** (які відрізняються тим, що ми в першому не знаємо дисперсію генеральної сукупності, але якщо в нас великий набір даних, то ці два тести дають дуже схожі результати) **ми перевіряємо, чи є різниця у середніх показниках по групам користувачів**.  

А в **тесті Хі-квадрат ми перевіряємо чи є звʼязок між групою користувача і тим, чи він зробить цікаву нам дію**. Це ніби дослідження одного і того самого, але дещо різними способами. Для перевірки, можна виконувати кілька тестів (особливо, якщо один дає якийсь непереконливий результат типу р-значення 0.07 - наче і fail to regect H0 на рівні стат значущості 5%, але цікаво, що скажуть інші тести), тож, зробимо і ми тест хі-квадрат та порівняємо його результат з z-тестом.

Про різницю між тестами можна почитати ще [тут](https://stats.stackexchange.com/a/178860) - це просто пояснення користувача стековерфлоу, але там розумні люди сидять.

Для проведення хі-квадрат тесту скористаємось функцією з `scipy.stats` `chi2_contingency` для обчислення статистики хі-квадрат і р-значення для перевірки конкретної гіпотези. У цю функцію вам треба передати таблицю 2х2: кількість випадків для кожної версії гри і значення `retention_7`.

**Задача**: виконайте тест хі-квадрат на рівні значущості 5% аби визначити, чи є залежність між версією гри та тим, чи зайде гравець на 7ий день після встановлення гри.
Тут гіпотези наступні
- Н0: значення retention_7 не залежить від версії гри
- Н1: є залежність між версією гри і значенням retention_7

Виведіть p-значення та зробіть висновок.


In [25]:
from scipy.stats import chi2_contingency

In [26]:
xi_table = pd.crosstab(df['version'], df['retention_7'])

xi_table

retention_7  False  True 
version                  
gate_30      36198   8502
gate_40      37210   8279

In [27]:
chi2, p_value_2, _, _ = chi2_contingency(xi_table)

print(f"Chi-square stat: {chi2:.2f}")
print(f"p-value: {p_value_2:.3f}")

# Висновки
if p_value_2 < 0.05:
    print("На рівні значущості 5% ми відхиляємо нульову гіпотезу: є залежність між версією гри і значенням retention_7.")
else:
    print("На рівні значущості 5% ми не можемо відкинути нульову гіпотезу: залежність між версією гри і значенням retention_7 не встановлена.")

Chi-square stat: 9.96
p-value: 0.002
На рівні значущості 5% ми відхиляємо нульову гіпотезу: є залежність між версією гри і значенням retention_7.
